agent/pkg/docker/docker.go

```diff
@@ -0,0 +1,471 @@
+}
+
+// ListRunningContainers lists running Docker containers satisfying the given filters.
+func (d *Client) ListRunningContainers(ctx context.Context, fs filters.Args) (
+	map[cproto.ID]types.Container, error,
+) {
+	// List "our" running containers, based on `dockerAgentLabel`.
+	// This doesn't affect fluentbit, or containers spawned by other agents.
+	containers, err := d.cl.ContainerList(ctx, types.ContainerListOptions{All: false, Filters: fs})
+	if err != nil {
+		return nil, err
+	}
+
+	result := make(map[cproto.ID]types.Container, len(containers))
+	for _, cont := range containers {
+		containerID, ok := cont.Labels[ContainerIDLabel]
+		if ok {
+			result[cproto.ID(containerID)] = cont
+		} else {
+			d.logger.Warnf("container %v has agent label but no container id", cont.ID)
+		}
+	}
+	return result, nil
+}
+
+// LabelFilter is a convenience that takes a key and value and returns a docker label filter.
+func LabelFilter(key, val string) filters.Args {
+	return filters.NewArgs(filters.Arg("label", key+"="+val))
+}
+
+func (d *Client) getDockerAuths(
+	ctx context.Context,
+	image reference.Named,
+	userRegistry *types.AuthConfig,
+	p Publisher,
+) (*types.AuthConfig, error) {
+	imageDomain := reference.Domain(image)
+	// Try user submitted registry auth config.
+	if userRegistry != nil {
+		// TODO: remove didNotPassServerAddress when it becomes required.
+		didNotPassServerAddress := userRegistry.ServerAddress == ""
+		if didNotPassServerAddress {
+			if err := p.Publish(ctx, NewLogEvent(model.LogLevelWarning,
+				"setting registry_auth without registry_auth.serveraddress is deprecated "+
+					"and will soon be required")); err != nil {
+				return nil, err
+			}
+		}
+
+		registryDomain := registry.ConvertToHostname(userRegistry.ServerAddress)
+		if registryDomain == imageDomain || didNotPassServerAddress {
+			return userRegistry, nil
+		}
+		if err := p.Publish(ctx, NewLogEvent(model.LogLevelWarning, fmt.Sprintf(
+			"not using expconfig registry_auth since expconf "+
+				"registry_auth.serverAddress %s did not match the image serverAddress %s",
+			registryDomain, imageDomain,
+		))); err != nil {
+			return nil, err
+		}
+	}
+
+	// Try using credential stores specified in ~/.docker/config.json.
+	if store, ok := d.credentialStores[imageDomain]; ok {
+		creds, err := store.get()
+		if err != nil {
+			return nil, fmt.Errorf("unable to get credentials from helper: %w", err)
+		}
+
+		if err := p.Publish(ctx, NewLogEvent(model.LogLevelInfo, fmt.Sprintf(
+			"domain '%s' found in 'credHelpers' config. Using credentials helper.",
+			imageDomain,
+		))); err != nil {
+			return nil, err
+		}
+
+		return &creds, nil
+	}
+
+	// Finally try using auths section of users ~/.docker/config.json.
+	index, err := registry.ParseSearchIndexInfo(image.String())
+	if err != nil {
+		return nil, fmt.Errorf("error invalid docker repo name: %w", err)
+	}
+	reg := registry.ResolveAuthConfig(d.authConfigs, index)
+	if reg != (types.AuthConfig{}) {
```

Comment:
Just want to confirm the desired logic here -- if `reg` _is_ empty, we still go ahead without returning an error, since that just means we should run without authentication?

Reply:
yep, that is the intended logic.

Reply:
refactored the flow a bit, to help (already refactored this entire function)